### DISCUSSION TITLE
v3

### DISCUSSION THREAD

**Discussion by brianzinn** [brianzinn/react-babylonjs#107] [General]
I have been absolutely swamped with work the last few weeks.  Lots of working till midnight and that will continue at least another week, but then v3 will be out sometime after that :)

There are currently 2 v-next branches. `v3` and `v3-proper`.  v3 was an experimental branch where `babylonjs-loaders` was a separate library that relied on another separate library for the Scene and Engine context.  I was hoping it would provide a path for a vanilla solution that didn't require this reconciler necessarily and the `babylonjs-loaders` project could be used on both paths.  It turned out to be a bit of a nightmare to work with - it required a lot of linking various projects and dependencies to work correctly and I think it would be a detriment to a clean developer experience and make it even harder to contribute, so abandoned that idea (and branch).  The other thing I wanted to do was on `useRef` to not have a `myRef.hostInstance`.  I made a lot of changes in the `v3` branch reconciler to support that and found that the fiber was not swapped when instances changed or moved around and additionally there is a built-in way on the reconciler 🤦  to provide the ref and still maintain the existing structure, which allows cleaner encapsulation and not polluting the babylon objects.  Additionally, I added in `v3-proper` some initial code that will (even declaratively) allow the babylon inspector to display custom metadata (and even actions) for react-babylonjs!  So, I need to bring in some bug fixes and also the code where I added ability to specify primary or secondary renderer (with a fix for re-using same renderer when scene changes, which logs errors in console from react).  I need to backport a few fixes from `v3` to `v3-proper` and it will be set for a new beta release.

TLDR: `v3-proper` will be the new v3 (ignore the `v3` branch).

**Comment by dennemark** [comment]
Thanks for the great wrap up of recent developments! Am looking forward to try out v3 in january! :)

**Comment by dennemark** [comment]
Hi Brian! Congrats to the recent update! I already read a bit through the recent changes and am curious to try out v3!
One thing I observed: The seperation of the api into its own page seems great, however it is currently missing remarks on hooks and model loaders. Maybe we can even wrap up this issue #87 if we add a description to the different model loaders. Hope to try out the assetmanager soon :) 

Btw. I will try to reply this week to the nodeMaterial hook #105 . I am wondering how I should contribute the hook. It seems the storybooks are still based on v2. Will you still seperate them from this respository?

**Comment by brianzinn** [comment]
@dennemark  You are right about docs on the hooks - completely missing!!  I will work on adding that this week (not Typedoc just manually for now).

For a node material hook - probably just add a new .ts file and export it and perhaps a storybook story showing usage.

As for storybook it will stay in this project for the foreseeable future as it's really easy to test changes without npm linking and keeps the storybooks up-to-date - in fact, all stories should all be working against v3.

**Comment by gauravarya83** [comment]
Hi Brian! congratulations and thanks for the latest update!
I have a question regarding useBabylonScene & useBabylonEngine hooks, seems these are removed in new version. Can you guide me how to get scene and engine in child components?

**Reply by brianzinn** [reply]
@gauravarya83 - thanks - those changes are a bit buried.  Need to navigate to the changes from the main page and there is a page just for the 3.0 breaking changes:
https://github.com/brianzinn/react-babylonjs/blob/master/docs/breaking-changes-2.x-to-3.0.md#hooks-renamed-usebabylonscene---usescene-usebabylonengine---useengine--usebabyloncanvas----usecanvas

I'm going to update the API docs as above with the hooks as well.  Let me know if you find anything else.  Cheers.

**Reply by gauravarya83** [reply]
Thanks @brianzinn and apologies that I did not refer docs upfront.
useScene and useEngine are working now, however, facing another issue (and this time followed the docs already :) ):

While using <Model... />, I am getting below error:
Unexpected token g in JSON at position 0 at babylonFileLoader.ts:499 (from sceneLoader.ts:744).

The only difference from the docs is that I am using <Model.../> in a child react component like:

App.js:
-------
return (
  <Engin...>
      <Scene...>
          <Suspense...>
            <Component1.../>
          <\/Suspense>
      <\/Scene>
  <\/Engine>
)

Component1.js:
----------------
return (
  <Box...>
      <Model.../>
  <\/Box>
)

**Reply by brianzinn** [reply]
yes, that's an important one that I missed @gauravarya83 !  Appreciate this feedback from you :)  I removed `@babylonjs/loaders` as a dependency entirely.  Previously `react-babylonjs` would register the loader for you and as now it's not a dependency anymore, so you need to load for side-effects.  How to do that is documented on the main babylon docs site here:
https://doc.babylonjs.com/divingDeeper/developWithBjs/treeShaking#loaders

The link above shows explicitly how to register only the loaders you need for your project.  That is very useful for reducing the size of your bundle (and was not possible in v2+).  I will update the breaking changes - thank-you for reporting.  The code in v2 that was called automatically was:
```javascript
// this will register loaders for all model types - likely you only need 1 or 2 in your project.
import '@babylonjs/loaders';
```

**Reply by gauravarya83** [reply]
Perfect, thanks. This is working now 👍 

I have another question (sorry I am new to this), is there a support for mesh.clone() in react-babylon, can you please point me to an example.
Basically, I have to clone the parent box (that I mentioned in previous example) after the Model is loaded.
Have tried in modelLoaded method, but it does not work: box.clone("box1");

PS: as per my understanding instancing does not clone children meshes.

**Reply by gauravarya83** [reply]
Just noticed that it compiled with useScene and useEngine, but both are returning null. Am I missing something?? :(

**Reply by brianzinn** [reply]
@gauravarya83  Are you using those hooks inside of the Scene component?

This won't work:
```jsx
const scene = useScene();

return (<Scene ... >... </Scene>)
```

You can only declare the useScene inside of the Scene.  It has to do with renderer boundaries, so this will work:
```jsx
const MySceneObjects = () => {
 const scene = useScene();
 console.log('scene:', scene);
 return (
   <arcRotateCamera .../>
   ... lights, meshes, etc
 )
}
const MyScene = () => (
  <Engine..>
    <Scene..>
      <MySceneObject />
    </Scene>
  </Engine>   
)
```

I'll make sure to include extra info in the API documentation for hooks.  It has to do with the context API and switching renderers and is a known issue with React.  Also, are you saying that it worked differently in 2.0 or you are starting fresh with 3.0?

^^ I should add that to address that I have considered moving away from Context API - it trips up a lot of people, because it is not that intuitive.

**Reply by gauravarya83** [reply]
Hi Brian, I am using it as you mentioned in the example, inside the child component. You can refer to my previous example, there I am using it inside Component1.js. Let me copy here again, and add the useScene statement:

App.js:
-------
return (
  <Engine...>
      <Scene...>
          <Suspense...>
            <Component1.../>
          <\/Suspense>
      <\/Scene>
  <\/Engine>
)

Component1.js:
----------------
import { useScene } from "react-babylonjs";

const scene = useScene();

return (
  <Box...>
      <Model.../>
  <\/Box>
)

It is working fine with useBabylonScene in 2.3 version.
I can share a public git tomorrow, if required.

**Reply by brianzinn** [reply]
@gauravarya83 I will definitely check this out as soon as I finish work!!  Thanks for the heads up.  If you can share a public git tomorrow then I can look also at the parentBox cloning.  Cheers.

**Reply by gauravarya83** [reply]
Surely, many thanks for your prompt responses :)

**Reply by brianzinn** [reply]
@gauravarya83 I just tried an example and it works:
```jsx
const JustBox = () => {
  const scene = useScene();
  console.log('scene:', scene);
  return <box size={2} name='box1' />
}
```
Thanks for all your feedback today - very helpful.  Create a public repo and github issue and we can work through them.  Cheers.

**Reply by gauravarya83** [reply]
Must be my bad then, let me investigate and put it in a public git, if it still does not work.

**Reply by gauravarya83** [reply]
Hi @brianzinn , I just investigated around this, and it is useEngine which is null not useScene.

Sharing git in a short while.

**Reply by gauravarya83** [reply]
Here is the git link: https://github.com/gauravarya83/playground
useEngine is null in lines 6, 7 of BoxedModel.js
To make it work for now, used const scene = useScene(); and then scene._engine...

I will now tryout the deep clone thing, will update you.

**Reply by brianzinn** [reply]
I agree that it should work as you have described as that is more intuitive.  I will roll out a new version - have just put the fix on master, but want to add another code change before NPM update.  thanks @gauravarya83 !! :)

**Comment by brianzinn** [comment]
hi @gauravarya83 you can do `mesh.clone()` yourself from a ref (it's a deep copy).  Better with an example (and I have not tried this) just typing a thought out:
```jsx
import { useSceneLoader, ...} from 'react-babylonjs';
import '@babylonjs/loaders';

const MyModel = ({ rootUrl, sceneFilename, parentBoxPosition }) => {
  const sceneLoaderResults = useSceneLoader(rootUrl, sceneFilename);
  
  useEffect(() => {
    return () => {
      // console.log('disposing the sceneloader results.')
      sceneLoaderResults.dispose();
    }
  }, []);

  const [parentBox, _] = useState(() => return sceneLoaderResults.meshMap['parent box']}
 
  return <>
      <abstractMesh fromInstance={sceneLoaderResults.rootMesh!} {...rest} />
      <mesh fromInstance={parentBox} position={parentBoxPosition} />
    </>
}
```

If that doesn't work - create an issue and we can work through it.  If you have a public mesh we can use that as an example.  Try to use the Scene Explorer to see if you have created the mesh - if it's in the same position as the original then it wouldn't be visible.

**Reply by gauravarya83** [reply]
Thanks for this Brian, I will try this and will update you.

**Reply by brianzinn** [reply]
hi  @gauravarya83   I don't know if you saw, but I updated the hooks to work as you expected in the new version.  Thanks for pointing it out.

**Reply by gauravarya83** [reply]
Many thanks @brianzinn , engine is back now 👍 
I am still struggling with cloning thing, just want to check it thoroughly before I come to you :)

**Reply by brianzinn** [reply]
@gauravarya83 no worries at all - update your github and I can look - just comment where you are stuck.  you've been really helpful with getting v3 out 😄

**Reply by gauravarya83** [reply]
Thanks @brianzinn I have updated the same git repo https://github.com/gauravarya83/playground
Here I have tried cloning/instancing in 3 components, please check Cloned, Cloned1 and Instanced.
You can uncomment the component you want to run in the App.js file

My objective is to have clones with working animations

PS: One challenge that I suspect is that this model has 6 meshes (one for each material inside the model and one for __root__)

**Reply by gauravarya83** [reply]
@brianzinn  please let me know.

**Reply by brianzinn** [reply]
hi @gauravarya83 sorry for the delay.  There is a playground example called `Instanced Bones`, which is what I used to change your Cloned page:
```jsx
import { useEffect } from "react";
import { Vector3 } from "@babylonjs/core";
import { useSceneLoader } from "react-babylonjs";
import "@babylonjs/loaders/glTF";

const Cloned = () => {
  const sceneLoaderResults = useSceneLoader("./", "Boy.glb");

  useEffect(() => {
    return () => {
      sceneLoaderResults.dispose();
    }
  }, [sceneLoaderResults]);

  return <>
    {[...Array(3).keys()].map((b, i) => <box key={i} name={`box${i}`} position={new Vector3(i * 2, -0.5, i * 2)}>
      {(i > 0) &&
        sceneLoaderResults.meshes.filter(mesh => mesh.subMeshes && mesh.subMeshes.length > 0).map(mesh => (
            <instancedMesh name={`${i}-${mesh.name}`} key={`${i}-${mesh.name}`} source={mesh} position={new Vector3(0, 0.5, 0)} />
          ))
      }
    </box>)}
  </>
}

export default Cloned;
```

One of your meshes has no geometry, so it can't be instanced, but looks to work okay.
![instanced-meshes](https://user-images.githubusercontent.com/542756/104998352-e89b4100-59df-11eb-844b-a4aad5f3e99d.gif)

There is something about the SceneLoader that it won't reload the same model.  I will probably need to write something to allow cache clearing.  You will see if you load the mesh and then try to load it again in the same scene.  Let me know how that goes.

**Reply by gauravarya83** [reply]
This is awesome, many thanks @brianzinn .. I will check further and let you know!!
